rama-net/src/fingerprint/peet/tls.rs

```diff
@@ -0,0 +1,375 @@
+use crate::tls::{
+    ApplicationProtocol, CertificateCompressionAlgorithm, CipherSuite, ExtensionId,
+    ProtocolVersion, SecureTransport, SignatureScheme, SupportedGroup,
+};
+use itertools::Itertools;
+use rama_core::context::Extensions;
+use std::fmt;
+
+#[derive(Clone)]
+/// Input data for a "peetprint" fingerprint.
+///
+/// Computed using a future `PeetPrint::compute` method.
+pub struct PeetPrint {
+    supported_tls_versions: Vec<ProtocolVersion>,
+    supported_protocols: Vec<ApplicationProtocol>,
+    supported_groups: Vec<SupportedGroup>,
+    supported_signature_algorithms: Vec<SignatureScheme>,
+    psk_key_exchange_mode: Option<u8>,
+    certificate_compression_algorithms: Option<Vec<CertificateCompressionAlgorithm>>,
+    cipher_suites: Vec<CipherSuite>,
+    sorted_extensions: Vec<String>,
+}
+
+impl PeetPrint {
+    pub fn compute(ext: &Extensions) -> Result<Self, PeetComputeError> {
+        let client_hello = ext
+            .get::<SecureTransport>()
+            .and_then(|st| st.client_hello())
+            .ok_or(PeetComputeError::MissingClientHello)?;
+        Self::compute_from_client_hello(client_hello)
+    }
+
+    pub fn compute_from_client_hello(
+        client_hello: impl ClientHelloProvider,
+    ) -> Result<Self, PeetComputeError> {
+        let cipher_suites: Vec<CipherSuite> = client_hello
+            .cipher_suites()
+            .filter(|c| !c.is_grease())
+            .collect();
+
+        if cipher_suites.is_empty() {
+            return Err(PeetComputeError::EmptyCipherSuites);
+        }
+
+        let mut supported_tls_versions = Vec::new();
+        let mut supported_protocols = Vec::new();
+        let mut supported_groups = Vec::new();
+        let mut supported_signature_algorithms = Vec::new();
+        let mut psk_key_exchange_mode = None;
+        let mut certificate_compression_algorithms = Vec::new();
+        let mut sorted_extensions = Vec::new();
+
+        for ext in client_hello.extensions() {
+            let id = ext.id();
+
+            if id.is_grease() {
+                sorted_extensions.push("GREASE".to_owned());
+                continue;
+            }
+
+            sorted_extensions.push(u16::from(id).to_string());
+
+            match ext {
+                ClientHelloExtension::SupportedVersions(versions) => {
+                    supported_tls_versions.extend(versions.iter().copied());
+                }
+                ClientHelloExtension::ApplicationLayerProtocolNegotiation(alpns) => {
+                    for p in alpns {
+                        match p {
+                            ApplicationProtocol::HTTP_10 => {}
+                            ApplicationProtocol::HTTP_11 => {}
+                            ApplicationProtocol::HTTP_2 => {}
+                            _ => {
+                                continue;
+                            }
+                        }
+
+                        supported_protocols.push(p.clone());
```

Comment:
I would already push write this as `&'static str` so that you do not need to validate afterwards anymore. So a `Vec<&'static str>` :) Or if you wanna optimise a bit this can probably be even a `SmallVec<[&'static str; 2]>` or so, as I don't think this will often be more than 2  :P but that's perhaps already a bit too overoptimised.